### PR TITLE
[RFR] Add trace logging for ssl-config

### PR DESCRIPTION
**Pull request by wsargent**
Fix for https://github.com/lightbend/ssl-config/issues/142

This is a logging proxy around SSLContext / SSLEngine / SSLSocket that is configured using `SSLDebugConfig` parameters as best it can.  This is based around the logging prototype described in https://tersesystems.com/blog/2018/10/20/jsse-with-object-capabilities/

### CODE REVIEW

**Comment by wsargent** [comment]
@dwijnand think this is ready for review now -- maybe 0.4.0 instead of 0.3.8?

**Comment by dwijnand** [comment]
Thanks, Will.  0.4.0 sounds fine if necessary.

I'm on holiday this week, out of the office next week, and then have 2 days where I'm on support rota.  So I'll _try_ to get to reviewing this this year, but no promises.

But I won't forget it as I want to unbreak master, so thank you for pushing on this.

If anyone else in @lightbend/play-lagom or @lightbend/akka-team is happy to review and land this, feel free to.

File: ssl-config-core/src/main/scala/com/typesafe/sslconfig/ssl/SSLContextBuilder.scala [hunk 1]
```diff
@@ -132,22 +134,37 @@ class ConfigSSLContextBuilder(
     builder.build()
   }
 
-  def buildCompositeKeyManager(keyManagerConfig: KeyManagerConfig, algorithmChecker: AlgorithmChecker) = {
+  @deprecated("Use newer version of method", "0.3.8")
```
Comment (marcospereira) [inline ssl-config-core/src/main/scala/com/typesafe/sslconfig/ssl/SSLContextBuilder.scala]:
```suggestion
  @deprecated("Use newer buildCompositeKeyManager with debug parameter", "0.3.8")
```

File: ssl-config-core/src/main/scala/com/typesafe/sslconfig/ssl/SSLContextBuilder.scala [hunk 2]
```diff
@@ -132,22 +134,37 @@ class ConfigSSLContextBuilder(
...
+  def buildCompositeKeyManager(keyManagerConfig: KeyManagerConfig, algorithmChecker: AlgorithmChecker, debug: SSLDebugConfig): CompositeX509KeyManager = {
     val keyManagers = keyManagerConfig.keyStoreConfigs.map {
       ksc =>
-        buildKeyManager(ksc, algorithmChecker)
+        buildKeyManager(ksc, algorithmChecker, debug)
     }
     new CompositeX509KeyManager(mkLogger, keyManagers)
   }
 
+  @deprecated("Use newer version of method", "0.3.8")
```
Comment (marcospereira) [inline ssl-config-core/src/main/scala/com/typesafe/sslconfig/ssl/SSLContextBuilder.scala]:
```suggestion
  @deprecated("Use newer version of buildCompositeTrustManager with debug parameter", "0.3.8")
```

File: ssl-config-core/src/main/scala/com/typesafe/sslconfig/ssl/SSLContextBuilder.scala [hunk 3]
```diff
@@ -200,10 +217,16 @@ class ConfigSSLContextBuilder(
   def warnOnPKCS12EmptyPasswordBug(ksc: KeyStoreConfig): Boolean =
     ksc.storeType.equalsIgnoreCase("pkcs12") && !ksc.password.exists(!_.isEmpty)
 
+  @deprecated("Use newer version of method", "0.3.8")
```
Comment (marcospereira) [inline ssl-config-core/src/main/scala/com/typesafe/sslconfig/ssl/SSLContextBuilder.scala]:
```suggestion
  @deprecated("Use newer version of buildKeyManager with debug parameter", "0.3.8")
```

File: ssl-config-core/src/main/resources/reference.conf [hunk 4]
```diff
@@ -131,60 +128,34 @@ ssl-config {
...
 
-    # Print trust manager tracing
+    # Enable trust manager tracing
     trustmanager = false
 
-    # Turn pluggability debugging on
-    pluggability = false
-
+    // The following settings are deprecated and have no effect in code.
+    certpath = false # DEPRECATED
```
Comment (marcospereira) [inline ssl-config-core/src/main/resources/reference.conf]: It would be good to have a warning logged when these are configured to `true`.

Comment (wsargent) [reply ssl-config-core/src/main/resources/reference.conf]: Added in https://github.com/lightbend/ssl-config/pull/143/commits/1660915e77c3e0eb0c41b9493fc42586c589808c

**Comment by dwijnand** [comment]
Needs a rebase now.

And I need to get some courage and try to review your two PRs...

**Comment by dwijnand** [comment]
Thanks, I might take you up on that offer.  I'll update the team and see if they want to join.

**Comment by SethTisue** [comment]
needs rebase

**Comment by wsargent** [comment]
@SethTisue ack

**Comment by wsargent** [comment]
@SethTisue `git rebase master` on branch showed a single test file that was deleted in the branch but updated in master.  Deleted and force pushed.

**Comment by dwijnand** [comment]
Given Marcos' review feedback and response, I'm going to merge this.

But I'm still concerned about the binary incompatibility, so I might restore some binary APIs (like no-op APIs).

**Comment by wsargent** [comment]
Woo!

**Pull request merged**